components/script/dom/element.rs

```diff
@@ -337,49 +332,63 @@ impl Element {
     }
 
     pub fn set_custom_element_state(&self, state: CustomElementState) {
-        self.custom_element_state.set(state);
+        self.ensure_rare_data().custom_element_state = state;
     }
 
     pub fn get_custom_element_state(&self) -> CustomElementState {
-        self.custom_element_state.get()
+        if let Some(rare_data) = self.rare_data().as_ref() {
+            return rare_data.custom_element_state;
+        }
+        CustomElementState::Undefined
     }
 
     pub fn set_custom_element_definition(&self, definition: Rc<CustomElementDefinition>) {
-        *self.custom_element_definition.borrow_mut() = Some(definition);
+        self.ensure_rare_data().custom_element_definition = Some(definition);
     }
 
     pub fn get_custom_element_definition(&self) -> Option<Rc<CustomElementDefinition>> {
-        (*self.custom_element_definition.borrow()).clone()
+        if let Some(rare_data) = self.rare_data().as_ref() {
+            return rare_data.custom_element_definition.clone();
+        }
+        None
     }
 
     pub fn push_callback_reaction(&self, function: Rc<Function>, args: Box<[Heap<JSVal>]>) {
-        self.custom_element_reaction_queue
-            .borrow_mut()
+        self.ensure_rare_data()
+            .custom_element_reaction_queue
             .push(CustomElementReaction::Callback(function, args));
     }
 
     pub fn push_upgrade_reaction(&self, definition: Rc<CustomElementDefinition>) {
-        self.custom_element_reaction_queue
-            .borrow_mut()
+        self.ensure_rare_data()
+            .custom_element_reaction_queue
             .push(CustomElementReaction::Upgrade(definition));
     }
 
     pub fn clear_reaction_queue(&self) {
-        self.custom_element_reaction_queue.borrow_mut().clear();
+        self.ensure_rare_data()
```

Comment:
This shouldn't call `ensure_rare_data`. If there's no rare data there's nothing to clear.

So `if let Some(data) = self.rare_data_mut() { data.custom_element_reaction_queue.clear() }` or such.